tests/integration/multithreading_test.py

```diff
@@ -50,3 +56,92 @@ def test_decorator_works_in_a_new_thread():
+    with zipkin_span(
+        service_name='_one_little_request',
+        span_name='do-the-thing',
+        binary_annotations={'input.was': my_input},
+    ) as span_ctx:
+        print('_do_one_little_request(%s): in new span; %r' % (
+            my_input, get_default_tracer()))
+        time.sleep(random.random())
+        span_ctx.add_annotation('time-' + my_input)
+        output_q.put(my_input + '-output')
+
+
+def _do_test_concurrent_subrequests_in_threads(thread_class):
+    """What if a thread has a span context, then fires off N threads to do N
+    subrequests in parallel?
+
+    Surely the spans of the subrequests should be siblings whose parentID is
+    equal to the first thread's span context's spanID.
+
+    Furthermore, a final sub-span created in the first thread should be a child
+    of that thread (no leaking of span stack from the child threads).
+    """
+    transport = MockTransportHandler()
+    with zipkin_span(
+        service_name='main_thread',
+        span_name='handle_one_big_request',
+        transport_handler=transport,
+        sample_rate=100.0,
+        encoding=Encoding.V2_JSON,
+        use_128bit_trace_id=True,
+    ) as span_ctx:
+        print('in root span; main thread: %r' % get_default_tracer())
+        assert True is span_ctx._is_local_root_span
+        assert span_ctx.logging_context
+        expected_trace_id = span_ctx.zipkin_attrs.trace_id
+        expected_parent_id = span_ctx.zipkin_attrs.span_id
+
+        # Now do three subrequests
+        req_count = 3
+        threads = []
+        output_q = queue.Queue()
+        for thread_idx in range(req_count):
+            this_thread = thread_class(
+                target=_do_one_little_request, args=(output_q,
+                                                     'input-%d' % thread_idx))
+            threads.append(this_thread)
+            this_thread.start()
+        outputs = set()
+        for thread in threads:
+            thread.join()
+            outputs.add(output_q.get())
+        assert {'input-0-output', 'input-1-output', 'input-2-output'} == outputs
+
+    output = transport.get_payloads()
+    assert len(output) == 1
+
+    spans = sorted(json.loads(output[0]), key=itemgetter('timestamp'))
+    assert len(spans) == 4
+    parent_span = spans[0]
+    subrequest_spans = spans[1:]
+    assert len(subrequest_spans) == 3
+    assert parent_span['name'] == 'handle_one_big_request'
+    assert [s['name'] for s in subrequest_spans] == ['do-the-thing'] * 3
+    assert set(s['tags']['input.was'] for s in subrequest_spans) == {
+        'input-0', 'input-1', 'input-2'}
+    assert [s['traceId'] for s in subrequest_spans] == [expected_trace_id] * 3
+    # Perhaps most importantly, all the subrequest spans should share the same
+    # parentId, which is the main thread's span's 'id'
+    assert [s['parentId'] for s in subrequest_spans] == [expected_parent_id] * 3
```

Comment:
can you just do a big for loop and put the asserts in there? I think it's more easily readable than all the list comprehensions

Reply:
yup!